spec/ruby/core/io/timeout_spec.rb

```diff
@@ -0,0 +1,201 @@
+
+ruby_version_is "3.2" do
+  describe "IO#timeout=" do
+    before :each do
+      @rpipe, @wpipe = IO.pipe
+    end
+
+    after :each do
+      @rpipe.close
+      @wpipe.close
+    end
+
+    it "can be set and read on File instances" do
+      fname = tmp("io_timeout_attribute.txt")
+      touch(fname)
+      file = File.open(fname, "a+")
+
+      file.timeout.should == nil
+
+      file.timeout = 1.23
+      file.timeout.should == 1.23
+    ensure
+      file.close
+      rm_r fname
+    end
+
+    it "can be set and read on IO instances" do
+      @rpipe.timeout.should == nil
+      @wpipe.timeout.should == nil
+
+      @rpipe.timeout = 1.23
+      @wpipe.timeout = 4.56
+
+      @rpipe.timeout.should == 1.23
+      @wpipe.timeout.should == 4.56
+    end
+
+    it "raises TypeError when incorrect value is provided for timeout" do
+      -> { @rpipe.timeout = -1 }.should raise_error(ArgumentError, 'time interval must not be negative')
+      -> { @rpipe.timeout = "1" }.should raise_error(TypeError, 'can\'t convert String into time interval')
+    end
+
+    it "accepts objects for interval that have the :divmod method defined" do
+      s = "abc"
+      s.define_singleton_method(:divmod) { |_| [0, 0.001] }
+
+      @rpipe.timeout = s
+      @rpipe.timeout.should == s
+    end
+  end
+
+  describe "IO#timeout" do
+    before :each do
+      @rpipe, @wpipe = IO.pipe
+      # There is no strict long term standard for pipe limits (2**16 bytes currently). This is an attempt to set a safe
+      # enough size to test a full pipe.
+      @more_than_pipe_limit = 1 << 18
+    end
+
+    after :each do
+      @rpipe.close
+      @wpipe.close
+    end
+
+    it "raises TypeError when incorrect value is observed for timeout at the time of use" do
+      klass = Class.new do
+        attr_accessor(:v)
+        def divmod(_); [v, 0]; end
+      end
+
+      o = klass.new
+      o.v = 1
+
+      @rpipe.timeout = o
+
+      o.v = -1
+      -> { @rpipe.read(1) }.should raise_error(ArgumentError, 'time interval must not be negative')
+    end
+
+    it "raises IO::TimeoutError when timeout is exceeded for .read" do
+      @rpipe.timeout = 0.001
+      -> { @rpipe.read.should }.should raise_error(IO::TimeoutError)
+    end
+
+    it "raises IO::TimeoutError when timeout is exceeded for .read(n)" do
+      @rpipe.timeout = 0.001
+      -> { @rpipe.read(3) }.should raise_error(IO::TimeoutError)
+    end
+
+    it "raises IO::TimeoutError when timeout is exceeded for .gets" do
+      @rpipe.timeout = 0.001
+      -> { @rpipe.gets }.should raise_error(IO::TimeoutError)
+    end
+
+    it "makes a last attempt to read when the time is over" do
+      @wpipe.write("abc")
+      @wpipe.close
+
+      @rpipe.timeout = 0
+      @rpipe.read(3).should == "abc"
+    end
+
+    it "waits for a read and completes" do
+      @rpipe.timeout = 0.1
+      read_result = nil
+
+      t = Thread.new do
+        read_result = @rpipe.read(3)
+      end
+
+      sleep(0.001)
```

Comment:
A rule of thumb is never used `sleep` in specs/tests, it's inherently causing transient behavior.
In this case you can use `Thread.pass until t.stop?`.
Same below of course.

Reply:
Got it. Done.